### ISSUE TITLE
Support workspace/didChangeWatchedFiles

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform-ls#15]
## Background

LSP method `textDocument/didChange` (which is already implemented) can be used to find out about changes made through the IDE, but sometimes files are changed outside of the IDE.

These can be detected through [`workspace/didChangeWatchedFiles`](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeWatchedFiles) which the LS doesn't implement yet.

Clients may choose to send pretty much arbitrary updates to the server and it is up to the server to filter them. Ideally though, the server should document what clients should watch so it doesn't have to filter ~garbage~ unnecessary updates.

Server may also _add_ to the list of watched globs, which is a feature described under https://github.com/hashicorp/terraform-ls/issues/867

## UX Impact

Users will see completion, go-to-definition or go-to-references etc. involving indexed but unopened modules to be accurate even when these modules change outside of the editor - e.g. when git branch is switched or files are changed in another editor.

## Proposal

 - [ ] Implement/register new handler for `workspace/didChangeWatchedFiles`
 - [ ] Filter files based on appropriate criteria
     - `*.tf` and `*.tfvars` files
     - directory is already tracked as a module, i.e. needs updating https://github.com/hashicorp/terraform-ls/blob/c277ec6e4daf7225fbcd655cd444a251ad3572d8/internal/langserver/handlers/did_change.go#L49-L53
  - [ ] Re-parse the directory in which relevant files were changed/added/delete https://github.com/hashicorp/terraform-ls/blob/c277ec6e4daf7225fbcd655cd444a251ad3572d8/internal/langserver/handlers/did_change.go#L55-L58
  - [ ] Document what files does the server expect to be watched by default in https://github.com/hashicorp/terraform-ls/blob/main/docs/language-clients.md

**Comment by radeksimko** [comment]
@danishprakash This might be relatively simple to resolve - or at least it doesn't have any further upstream dependencies, if you want to tackle a new issue.

**Comment by danishprakash** [comment]
Hey @radeksimko, I've a few questions while going through this issue, namely:
1. The spec mentions that the file events can include folders as well. In this case, do we use "modules" or do we walk through the directory and update each file there. I'm still not very clear as to what a module is in the server's context (is it the initialized directory generated by `terraform init`?), wasn't able to find much to read around this in the code.
2. Regarding the registration of the watchers, how are we handling registration for other methods? For instance, how is the registration for `DidChangeTextDocument` currently handled? Didn't find any references around this in the codebase at least(I might be missing something though). Otoh, I also noticed that `gopls` for instance, has some of these glob patterns hard-coded in the server. But at least from the spec and your comment, it seems like it's the client who should define glob patterns for the files to be watched.

Apologies if these are rather trivial questions, I thought It better to clear them before starting with the implementation.

**Comment by danishprakash** [comment]
Sorry about the delay on this @radeksimko, got caught up with work.

Thanks for the thorough response to my questions. I'm still trying to understand the whole thing and from what I can gather till now, we're essentially handling the notifications here. When I mentioned "registration", I actually meant the server registration to the client for a specific event for e.g. [this](https://github.com/golang/tools/blob/master/internal/lsp/general.go#L361-L372) is how gopls registers for the `didChangeWatchedFiles` event from the server (to the client) and I'm assuming or missing that we're not handling the registration for our handlers. Please correct me here.

With that being said, I'm thinking of moving ahead with [this](https://github.com/hashicorp/terraform-ls/compare/main...danishprakash:danish-didchangewatchedfiles#diff-da239365cad9dbf57a342a8b8cc8a7ce73e1d5a9eef220ffbbeb9d60c55ac651R32-R44) outline. I can go ahead and raise a WIP PR where you can also track the progress if that looks like the right direction.

**Comment by danishprakash** [comment]
Sorry about the delay on this @radeksimko.

Thanks for the thorough response to my questions. I'm still trying to understand the whole thing and from what I can gather till now, we're essentially handling the notifications here. When I mentioned "registration", I actually meant the server registration to the client for a specific event for e.g. [this](https://github.com/golang/tools/blob/master/internal/lsp/general.go#L361-L372) is how gopls registers for the `didChangeWatchedFiles` event from the server (to the client) and I'm assuming or missing that we're not handling the registration for our handlers. Please correct me here.

With that being said, I'm thinking of moving ahead with [this](https://github.com/hashicorp/terraform-ls/compare/main...danishprakash:danish-didchangewatchedfiles#diff-da239365cad9dbf57a342a8b8cc8a7ce73e1d5a9eef220ffbbeb9d60c55ac651R32-R44) outline. I can go ahead and raise a WIP PR where you can also track the progress if that looks like the right direction.

**Comment by radeksimko** [comment]
> I actually meant the server registration to the client for a specific event for e.g. this is how gopls registers for the didChangeWatchedFiles event from the server (to the client) and I'm assuming or missing that we're not handling the registration for our handlers.

That's correct - dynamic capability registration is an AFAIK optional part of LSP which our server _currently_ doesn't support. This means that all capability negotiation has to happen early during `initialize` request/response and nothing can change afterwards without restarting the server.

There is no reason for not supporting dynamic (un)registration, other than just "there are other tasks with higher priority currently".

> I'm thinking of moving ahead with this outline. I can go ahead and raise a WIP PR where you can also track the progress if that looks like the right direction.

That looks like a reasonable start, feel free to raise a draft PR.

**Comment by radeksimko** [comment]
FYI @danishprakash I have clarified the scope and added some more notes to the initial comment here + I also created https://github.com/hashicorp/vscode-terraform/issues/1068 which may be an easy way to test this.

Let me know if you need any further help with https://github.com/hashicorp/terraform-ls/pull/790 or if you'd prefer for anyone else to take it over. We'll likely need to prioritize it in the upcoming weeks as part of larger list tracked in https://github.com/hashicorp/vscode-terraform/issues/715

**Comment by danishprakash** [comment]
@radeksimko thanks for the heads up, I looked into it last week and I'll share an update by EOD today.
